### PR TITLE
Added volumetric fog effect.

### PR DESCRIPTION
**Pull request by reduz**
Add a customizable volumetric fog based on: 
https://bartwronski.files.wordpress.com/2014/08/bwronski_volumetric_fog_siggraph2014.pdf

Lights can get added to the fog volume to create shafts, godrays, etc. The effect is more or less subtle (its worth investigating temporal reprojection of fog volumes to increase sharpness). 

Additionally, both GI systems in Godot (SDFGI and GIProbe) can inject their light into the fog volume, for pretty volumetric GI.

image
image
image

### CODE REVIEW

File: servers/rendering/rasterizer.h [hunk 1]
```diff
@@ -88,6 +88,13 @@ class RasterizerScene {
 	virtual void environment_glow_set_use_bicubic_upscale(bool p_enable) = 0;
 	virtual void environment_set_fog(RID p_env, bool p_enable, float p_begin, float p_end, RID p_gradient_texture) = 0;
 
+	virtual void environment_set_volumetric_fog(RID p_env, bool p_enable, float p_density, const Color &p_light, float p_light_energy, float p_lenght, float p_detail_spread, float p_gi_inject, RS::EnvVolumetricFogShadowFilter p_shadow_filter) = 0;
+
+	virtual void environment_set_volumetric_fog_volume_size(int, int) = 0;
```
Comment (clayjohn) [inline servers/rendering/rasterizer.h:93]: I guess add names to the parameters

**Review by clayjohn — COMMENTED** [review]
I only had time for a brief review tonight, but it looks good so far! I noticed that fog isn't being applied over the sky, is there a setting for that that I missed, or did you forget to add it?

Its a lot of fun playing around with  and the sdfgi!

Probably also a good time to review https://github.com/godotengine/godot/pull/36989

File: servers/rendering/rasterizer.h (hunk 1 above)
Comment (AnilBK) [reply servers/rendering/rasterizer.h]: Small typo:
p_lenght instead of p_length

**Comment by Zireael07** [comment]
Vulkan only or GLES3 too?

**Comment by LinuxUserGD** [comment]
The GLES2/GLES3 code doesn't exist anymore after https://github.com/godotengine/godot/pull/41219 , so I guess Vulkan only for now? @Zireael07

File: servers/rendering/rasterizer.h (hunk 1 above)
Comment (Sonoracpp) [reply servers/rendering/rasterizer.h]: Should be p_length instead of p_lenght

**Comment by Zireael07** [comment]
@LinuxUserGD: I know GLES2 doesn't exist currently, I was asking about whether the effect can be backported later to GLES**3** as some of the graphical improvements are Vulkan only (and a lower-end backend will eventually be needed since not everyone has Vulkan, especially on mobile devices)

**Pull request merged**